src/generate/register.rs

```diff
@@ -1400,3 +1474,83 @@ fn lookup_filter(
         .find(|evsbase| evsbase.0.usage == Some(usage))
         .or_else(|| evs.first())
 }
+
+fn enums_to_map(evs: &EnumeratedValues) -> BTreeMap<u64, &EnumeratedValue> {
+    let mut map = BTreeMap::new();
+    for ev in &evs.values {
+        if let Some(v) = ev.value {
+            map.insert(v, ev);
+        }
+    }
+    map
+}
+
+fn minimal_hole(map: &BTreeMap<u64, &EnumeratedValue>, width: u32) -> Option<u64> {
+    (0..(1u64 << width)).find(|&v| !map.contains_key(&v))
+}
+
+fn add_from_variants_with_default(
+    mod_items: &mut TokenStream,
+    variants: &[Variant],
+    default: &Variant,
+    pc: &Ident,
+    fty: &Ident,
+    desc: &str,
+    reset_value: Option<u64>,
+) {
+    let repr = if fty == "bool" {
+        quote!()
+    } else {
+        quote! { #[repr(#fty)] }
+    };
+
+    let mut vars = TokenStream::new();
+    let mut casts = TokenStream::new();
+    for (v, c) in variants.iter().chain(std::iter::once(default)).map(|v| {
+        let desc = util::escape_special_chars(&util::respace(&format!("{}: {}", v.value, v.doc)));
+        let pcv = &v.pc;
+        let pcval = &util::unsuffixed(v.value);
+        (
+            quote! {
+                #[doc = #desc]
+                #pcv,
+            },
+            quote! {
+                #pc::#pcv => #pcval,
+            },
+        )
+    }) {
+        vars.extend(v);
+        casts.extend(c);
+    }
```

Comment:
this seems wrong, the repr is wrong now when compared to the actual value on chip

```suggestion
    for (v, c) in variants.iter().chain(std::iter::once(default)).map(|v| {
        let desc = util::escape_special_chars(&util::respace(&format!("{}: {}", v.value, v.doc)));
        let pcv = &v.pc;
        let pcval = &util::unsuffixed(v.value);
        (
            quote! {
                #[doc = #desc]
                #pcv = #pcval,
            },
            quote! {
                #pc::#pcv => #pcval,
            },
        )
    }) {
        vars.extend(v);
        casts.extend(c);
    }
```